[ derive ] Support data specialisation

UARTman opened the pull request.

buzden committed:
[ test ] Add tests with a need for specialisation

buzden committed:
[ test ] Add tests with a need for specialisation

UARTman committed:
[ refactor ] Rename specialisation functions

UARTman committed:
[ refactor ] Rename specialisation functions

UARTman committed:
[ new ] Implement normaliseTask

UARTman committed:
[ new ] Implement normaliseTask

UARTman committed:
[ new ] Sketch out specialiseIfNeeded

UARTman committed:
[ new ] Sketch out specialiseIfNeeded

UARTman committed:
[ cleanup ] Remove unneeded lets from specialiseIfNeeded

UARTman committed:
[ cleanup ] Remove unneeded lets from specialiseIfNeeded

UARTman committed:
[ refactor ] Move specialisation-related code to Util.Specialisation

UARTman committed:
[ refactor ] Move specialisation-related code to Util.Specialisation

UARTman committed:
[refactor, new] Add first test for specialiseIfNeeded

UARTman committed:
[refactor, new] Add first test for specialiseIfNeeded

UARTman committed:
[ fix ] Remove some type ambiguity from decEq implementation derivator

UARTman committed:
[ fix ] Remove some type ambiguity from decEq implementation derivator

UARTman committed:
[ new ] Overhaul `processArgs` for more correct specialisation task generation

UARTman committed:
[ new ] Overhaul `processArgs` for more correct specialisation task generation

UARTman committed:
[ new ] Make `processArgs` more aware of application semantics

UARTman committed:
[ new ] Make `processArgs` more aware of application semantics

UARTman committed:
[ cleanup ] Break up overlong line

UARTman committed:
[ cleanup ] Break up overlong line

UARTman committed:
[ cleanup ] Minor cleanup

UARTman committed:
[ cleanup ] Minor cleanup

UARTman committed:
[ fix ] Use different awk invocation

UARTman committed:
[ fix ] Use different awk invocation

UARTman committed:
[ new ] Autogenerate specialised type names using hashes of specialisation tasks

UARTman committed:
[ new ] Autogenerate specialised type names using hashes of specialisation tasks

UARTman committed:
[ fix ] Use local names in derived specialised types

UARTman committed:
[ fix ] Use local names in derived specialised types

UARTman committed:
[ fix ] Allow detection of existing specialised types in local context

UARTman committed:
[ fix ] Allow detection of existing specialised types in local context

UARTman committed:
[ cleanup ] Unpin hashable-derive

UARTman committed:
[ cleanup ] Unpin hashable-derive

buzden committed:
[ derive ] Call to specialiser during derivation process

buzden committed:
[ derive ] Call to specialiser during derivation process

UARTman committed:
[ fix ] Don't specialise types with private constructors + immediately declare specialised types

UARTman committed:
[ fix ] Don't specialise types with private constructors + immediately declare specialised types

UARTman committed:
[ fix, refactor ] Fix `specialiseIfNeeded`

- Document `AllApps` and its methods
- Introduce `GenArg` instead of a tuple
- Document steps of `specialiseIfNeeded`
- Check if a type argument with given value is passthrough (if so, it doesn't count towards deciding if specialisation is necessary)
- Wrap specialised generator call in `map cast`

UARTman committed:
[ fix, refactor ] Fix `specialiseIfNeeded`

- Document `AllApps` and its methods
- Introduce `GenArg` instead of a tuple
- Document steps of `specialiseIfNeeded`
- Check if a type argument with given value is passthrough (if so, it doesn't count towards deciding if specialisation is necessary)
- Wrap specialised generator call in `map cast`

UARTman committed:
[ fix ] Make `.cleanup-names` nix-friendly

UARTman committed:
[ fix ] Make `.cleanup-names` nix-friendly

UARTman committed:
[ new ] Add logging for derived spec types

UARTman committed:
[ new ] Add logging for derived spec types

UARTman committed:
[ fix ] Hide SnocVect in le/r/gadt/015 test

UARTman committed:
[ fix ] Hide SnocVect in le/r/gadt/015 test

UARTman committed:
[ fix ] Update expected logs for specIfNeeded

UARTman committed:
[ fix ] Update expected logs for specIfNeeded

UARTman committed:
[ fix ] Update expecteds in derivation/core

UARTman committed:
[ fix ] Update expecteds in derivation/core

UARTman committed:
[ fix ] Add missing check to processArg

UARTman committed:
[ fix ] Add missing check to processArg

UARTman committed:
[ fix, new ] Add argument analysis logging and fix specIfNeeded tests

UARTman committed:
[ fix, new ] Add argument analysis logging and fix specIfNeeded tests

UARTman committed:
[ cleanup ] Fix editorconfig lint

UARTman committed:
[ cleanup ] Fix editorconfig lint

UARTman committed:
[ fix ] Remove some of the memory-explosion effects of logging

UARTman committed:
[ fix ] Remove some of the memory-explosion effects of logging

UARTman committed:
[ new ] Add trivial test on detecting type arg passthroughs

UARTman committed:
[ new ] Add trivial test on detecting type arg passthroughs

UARTman committed:
[ new ] Implement specialised constructor name erasure

UARTman committed:
[ new ] Implement specialised constructor name erasure

UARTman committed:
[ fix ] Prefix name-erased constructors with typename

UARTman committed:
[ fix ] Prefix name-erased constructors with typename

UARTman committed:
[ new ] Add test for deriving spec list generators

UARTman committed:
[ new ] Add test for deriving spec list generators

UARTman committed:
[ new ] Add test for spec vects

UARTman committed:
[ new ] Add test for spec vects

UARTman committed:
[ new ] Add test for Vect n (Fin n) derivation

UARTman committed:
[ new ] Add test for Vect n (Fin n) derivation

UARTman committed:
[ delete ] Remove redundant spec tests from core

UARTman committed:
[ delete ] Remove redundant spec tests from core

buzden committed:
[ test ] Add couple more tests for simple specialisation

buzden committed:
[ test ] Add couple more tests for simple specialisation

UARTman committed:
[ fix ] Fix cast-wrapping in specialised generators

UARTman committed:
[ fix ] Fix cast-wrapping in specialised generators

UARTman committed:
[ fix ] Make spec tests compile and fail only on totality check

UARTman committed:
[ fix ] Make spec tests compile and fail only on totality check

UARTman committed:
[ new ] Display specialised name in normal logs

UARTman committed:
[ new ] Display specialised name in normal logs

buzden committed:
[ test ] Add some lacking print tests + fix one run test

buzden committed:
[ test ] Add some lacking print tests + fix one run test

UARTman committed:
[ fix ] Fix incorrect gen signature in tests

UARTman committed:
[ fix ] Fix incorrect gen signature in tests

UARTman committed:
[ fix ] Fix bugs in specialiseIfNeeded logic

- Specialise on primitive types
- Use try-check for more correct visibility checking
- Generate fully qulaify the type's namespace when generating it

UARTman committed:
[ fix ] Fix bugs in specialiseIfNeeded logic

- Specialise on primitive types
- Use try-check for more correct visibility checking
- Generate fully qulaify the type's namespace when generating it

UARTman committed:
[ test ] Add expected values to succeeding tests

UARTman committed:
[ test ] Add expected values to succeeding tests

UARTman committed:
[ cleanup ] Break up too-log lines

UARTman committed:
[ cleanup ] Break up too-log lines

UARTman committed:
[ test ] Update specialise-if-needed outputs

UARTman committed:
[ test ] Update specialise-if-needed outputs

UARTman committed:
[ test ] Fix more tests

UARTman committed:
[ test ] Fix more tests

UARTman committed:
[ test ] Fixup print tests

UARTman committed:
[ test ] Fixup print tests

UARTman committed:
[ test ] Add test for obscure spec bug (yet unfixed)

UARTman committed:
[ test ] Add test for obscure spec bug (yet unfixed)

UARTman committed:
[ fix ] Prepend type name to type arguments

UARTman committed:
[ fix ] Prepend type name to type arguments

UARTman committed:
[ test ] Port specialise-data tests over to symlinks

UARTman committed:
[ test ] Port specialise-data tests over to symlinks

UARTman committed:
[ fix ] Allow for dependent lambdas in specialisation

UARTman committed:
[ fix ] Allow for dependent lambdas in specialisation

UARTman committed:
[ test ] Adjust spec-if-needed expecteds for 2e969a89

UARTman committed:
[ test ] Adjust spec-if-needed expecteds for 2e969a89

UARTman committed:
[ new ] Sketch out the recursive constructor argument finder

UARTman committed:
[ new ] Sketch out the recursive constructor argument finder

UARTman committed:
[ new ] Generate separate specialised type signature

UARTman committed:
[ new ] Generate separate specialised type signature

UARTman committed:
[ refactor ] Pre-cache specialised type invocation

UARTman committed:
[ refactor ] Pre-cache specialised type invocation

UARTman committed:
[ refactor ] Decouple claims from definitions

UARTman committed:
[ refactor ] Decouple claims from definitions

UARTman committed:
[ new ] Implement recursive argument  checking

- breaks function derivation in recursive spec types

UARTman committed:
[ new ] Implement recursive argument  checking

- breaks function derivation in recursive spec types

UARTman committed:
[ refactor ] Add type metadata records, update packdb

UARTman committed:
[ refactor ] Add type metadata records, update packdb

UARTman committed:
[ new ] Implement new cast injectivity prover (WIP)

UARTman committed:
[ new ] Implement new cast injectivity prover (WIP)

UARTman committed:
[ new ] Give mToPImpl implicit arguments in cast injectivity prover

UARTman committed:
[ new ] Give mToPImpl implicit arguments in cast injectivity prover

UARTman committed:
[ new ] Make all specialisation tests pass

UARTman committed:
[ new ] Make all specialisation tests pass

UARTman committed:
[ new ] Fix long line lint

UARTman committed:
[ new ] Fix long line lint

UARTman committed:
[ refactor ] Clean up specialiser code

- Remove obsolete implementations and unneeded functions
- Document some new functions

UARTman committed:
[ refactor ] Clean up specialiser code

- Remove obsolete implementations and unneeded functions
- Document some new functions

UARTman committed:
[ new ] Add test for failing specialisation in pil-reg

It fails for now

UARTman committed:
[ new ] Add test for failing specialisation in pil-reg

It fails for now

UARTman committed:
[ new ] Improve normaliseTask type resolution

UARTman committed:
[ new ] Improve normaliseTask type resolution

UARTman committed:
[ refactor ] Adjust specdata-027 to be abour normalisation

UARTman committed:
[ refactor ] Adjust specdata-027 to be abour normalisation

UARTman committed:
[ fix ] Arg name <-> hole conversion in task normalisation

UARTman committed:
[ fix ] Arg name <-> hole conversion in task normalisation

UARTman committed:
[ fix ] Give normalizer better types

UARTman committed:
[ fix ] Give normalizer better types

UARTman committed:
[ new ] Use polymorphic types in PilFun

UARTman committed:
[ new ] Use polymorphic types in PilFun

UARTman committed:
[ fix ] Update pil-fun expected test

UARTman committed:
[ fix ] Update pil-fun expected test

UARTman committed:
[ fix ] Work around mgn quirks

UARTman committed:
[ fix ] Work around mgn quirks

UARTman committed:
[ fix ] Generate correct DPair sequences for specialiseds with >=2 givens

UARTman committed:
[ fix ] Generate correct DPair sequences for specialiseds with >=2 givens

UARTman committed:
[ new ] Use polymorphic types in PilDyn

UARTman committed:
[ new ] Use polymorphic types in PilDyn

UARTman committed:
[ fix ] Add ConsRecs to specialiseIfNeeded

UARTman committed:
[ fix ] Add ConsRecs to specialiseIfNeeded

UARTman committed:
[ fix ] Fix tests to work after rebase

UARTman committed:
[ fix ] Fix tests to work after rebase

UARTman committed:
[ fix ] Update specialisation logic in dtc following rebase

UARTman committed:
[ fix ] Update specialisation logic in dtc following rebase

UARTman committed:
[ perf ] Monadize another bottleneck
~5 times faster now

UARTman committed:
[ perf ] Monadize another bottleneck
~5 times faster now

UARTman committed:
[ fix ] Fix tests broken by recent rebase

UARTman committed:
[ fix ] Fix tests broken by recent rebase

UARTman committed:
[ clean ] Cleanup specialiser code

UARTman committed:
[ clean ] Cleanup specialiser code

UARTman committed:
[ clean ] Continue specialiser cleanup

UARTman committed:
[ clean ] Continue specialiser cleanup

UARTman committed:
[ clean ] Use AppApps from elab-util-extra

UARTman committed:
[ clean ] Revert TypeInfo.decl in favor of .declNoNS in SpecialiseData.idr

UARTman committed:
[ upstream ] Adapt specialisation print tests to upstream pretty-printer changes

UARTman committed:
[ cleanup ] Few cosmetic fixes

buzden commented on src/Deriving/DepTyCheck/Gen/ForAllNeededTypes/Impl.idr:
```diff
@@ -69,9 +73,14 @@ DeriveBodyForType => ClosuringContext m => Elaboration m => SortedMap GenSignatu
+
+    sinRet : Maybe TTImp <- if notfound then assert_total specialiseIfNeeded sig fuel values else pure Nothing
+    let Nothing = sinRet
+      | Just me => pure (me, Nothing)
```
Maybe, no variable should be there, just pattern matching on the left of `<-`

buzden commented on src/Deriving/DepTyCheck/Gen/ForAllNeededTypes/Impl.idr:
```diff
@@ -69,9 +73,14 @@ DeriveBodyForType => ClosuringContext m => Elaboration m => SortedMap GenSignatu
     -- check if internal generator asked for is for a primitive type
     when (isTypeInfoPrim sig.targetType) $
       fail "Cannot derive generator for the primitive type \{show $ extractTargetTyExpr sig.targetType}, use external instead"
+    notfound <- pure $ id $ not $ List.Set.contains sig %search
```
Please add a comment on why `pure $ id $ ...` is needed here

buzden commented on src/Deriving/DepTyCheck/Gen/ForAllNeededTypes/Impl.idr:
```diff
@@ -5,12 +5,16 @@ import public Control.Monad.State
 
 import public Data.DPair
 import public Data.List.Set
+import public Data.List.Ex
```
Please preserve the alphanumeric order of imports

buzden commented on src/Deriving/DepTyCheck/Util/Specialisation.idr:
```diff
@@ -0,0 +1,307 @@
+inSameNS (NS ns _) n = NS ns n
+inSameNS _ n = n
+
+export %tcinline
```
Are you sure such a big function needs to have `%tcinline`?

buzden commented on src/Deriving/DepTyCheck/Util/Specialisation.idr:
```diff
@@ -0,0 +1,307 @@
+|||
+||| (x -> (y -> z) -> q) becomes (? -> (? -> ?) -> ?)
+allQuestions : NamesInfoInTypes => TTImp -> TTImp
+allQuestions t = runIdentity $ mapATTImp' allQImpl t
```
It seems like `allQImpl` should be of signature `... => TTImp -> TTImp -> m TTImp` and `mapMTTImp'` should be used instead of `mapATTimp'`. This makes `allQImpl` much simpler since it can pattern match on the second argument directly instead of double pattern-matching. After than, there is no need to `Monad m` requirement, only for `Applicative m`.

buzden commented on src/Deriving/DepTyCheck/Util/Specialisation.idr:
```diff
@@ -0,0 +1,307 @@
+
+LogPosition GenArg where
+  logPosition (MkGenArg a Nothing) = "\{fromMaybe "" a.name}"
+  logPosition (MkGenArg a $ Just t) = "(\{fromMaybe "" a.name} := \{show t})"
```
Maybe an empty string is not the best default value. Consider, e.g., `"?"` or whatever non-empty

buzden commented on src/Deriving/DepTyCheck/Util/Specialisation.idr:
```diff
@@ -0,0 +1,307 @@
+  let givenIdxVals = Prelude.toList sig.givenParams `zipV` givenParamValues
+  let genArgs = mkArgs sig (withIndex sig.targetType.args) givenIdxVals
+  -- Check if at least one `GenArg` can be specialised upon (i.e. is a type argument and has a non-passthrough given value)
+  specable <- traverse (.isSpecialising) genArgs
```
A comment should be added that if this check isn't performed, then we can loop forever, since specialised data type would be equal to the original one.

buzden commented on src/Deriving/DepTyCheck/Util/Specialisation.idr:
```diff
@@ -0,0 +1,307 @@
+||| Determine if the argument should be specialised or passed through
+(.isSpecialising) : Elaboration m => GenArg -> m Bool
+(.isSpecialising) (MkGenArg a Nothing) = pure False
+(.isSpecialising) (MkGenArg a $ Just g) = do
```
We discussed and decided to rename it to `isPassThrough`, in this case the whole logic seems to be clearer

buzden commented on src/Deriving/DepTyCheck/Util/Specialisation.idr:
```diff
@@ -0,0 +1,307 @@
+    | _ => Nothing
+  Just $ map (uncurry MkGenArg) $ zip tyInfo.args $ popArgVals tyInfo.args (mkAllApps aTerms)
+
+||| Assemble a list of arguments and their given values from `callGen` inputs
```
I think it's important to state in the docstring that this function requires indices inside both given lists to be in ascending order

buzden commented on src/Deriving/DepTyCheck/Util/Specialisation.idr:
```diff
@@ -0,0 +1,307 @@
+  m (TTImp, List Arg, List $ Maybe TTImp)
+processArgs sig ga = map (bimap (reAppAny $ IVar EmptyFC sig.targetType.name) unGA) $ processArgs' sig 0 ga
+
+export
```
Please add a comment of what this function is meant to do, since it's exported

buzden commented on src/Deriving/DepTyCheck/Util/Specialisation.idr:
```diff
@@ -0,0 +1,307 @@
+
+allConstructorsVisible : Elaboration m => TypeInfo -> m Bool
+allConstructorsVisible ti = do
+  all id <$> traverse nameUnambigAndVis (name <$> ti.cons)
```
```suggestion
  all id <$> traverse (nameUnambigAndVis . name) ti.cons
```

buzden replied:
Just as a possible variant:
```suggestion
  let _ = Monoid.All in foldMap (nameUnambigAndVis . name) ti.cons
```

buzden commented on src/Deriving/DepTyCheck/Util/Specialisation.idr:
```diff
@@ -0,0 +1,307 @@
+dPairHelper 0 = `(?)
+dPairHelper (S n) = `(DPair ? $ \_ => ~(dPairHelper n))
+
+inSameNS : Name -> Name -> Name
```
In order to have self-documenting code

```suggestion
inSameNS : (nsSource : Name) -> Name -> Name
```

buzden commented on src/Deriving/DepTyCheck/Util/Specialisation.idr:
```diff
@@ -0,0 +1,307 @@
+
+dPairHelper : Nat -> TTImp
+dPairHelper 0 = `(?)
+dPairHelper (S n) = `(DPair ? $ \_ => ~(dPairHelper n))
```
I'd suggest renaming to something like `dPairOfUnknowns`

buzden commented on src/Deriving/DepTyCheck/Util/Specialisation.idr:
```diff
@@ -0,0 +1,307 @@
+    | False => do
+      logPoint DetailedDebug "deptycheck.util.specialisation" [sig]
+        "Not found any type arguments that can be specialised upon, specialisation impossible."
+      pure Nothing
```
```suggestion
      logValue DetailedDebug "deptycheck.util.specialisation" [sig]
        "Not found any type arguments that can be specialised upon, specialisation impossible."
        Nothing
```

buzden replied:
There are quite a few places below and above where this suggestion applies

buzden commented on src/Deriving/DepTyCheck/Util/Specialisation.idr:
```diff
@@ -0,0 +1,307 @@
+  let genArgs = mkArgs sig (withIndex sig.targetType.args) givenIdxVals
+  -- Check if at least one `GenArg` can be specialised upon (i.e. is a type argument and has a non-passthrough given value)
+  specable <- traverse (.isSpecialising) genArgs
+  let True = any id specable
```
Maybe, `specable` variable should be inlined with `!(...)` here

UARTman committed:
[ cleanup ] Address comments

UARTman committed:
[ fix ] Add [noHints] to specialiser options

UARTman committed:
[ refactor ] rename isSpecialising

UARTman committed:
[ cleanup ] Address more comments

UARTman committed:
[ fix ] Fix specialiser test 024

UARTman committed:
[ refactor ] Adopt logValue where possible

buzden commented on src/Deriving/DepTyCheck/Util/Specialisation.idr:
```diff
@@ -0,0 +1,307 @@
+    | _ => pure `(?)
+  case procL of
+    Implicit _ _ => pure `(?)
+    _ => pure r
```
```suggestion
allQImpl (IApp {}) (IApp _ (Implicit {}) _) = pure `(?)
allQImpl (IApp {}) r@(IApp {}) = pure r
allQImpl (IApp {}) _ = pure `(?)
```

buzden commented on elab-util-extra/src/Deriving/SpecialiseData.idr:
```diff
@@ -214,11 +208,13 @@ makeTypeArgM0 a = { count := if a.type == `(Type) then M0 else a.count } a
-mergeAliases m = mergeWith (Prelude.curry fst) m . fromList . map (mapSnd var)
+||| Assemble a TTImp of a tuple from a list of `TTImp`s
+tupleOf : List TTImp -> TTImp
+tupleOf [] = `(())
```
This syntax will lead to returning `IAlternative ...` between `Unit` and `MkUnit`. This *potentially* may cause longer compilation time for derived code, and *potentially* some compilation/ambiguity errors. I think what was meant is this:
```suggestion
tupleOf [] = `(MkUnit)
```

buzden commented on examples/pil-fun/src/Language/PilFun/Pretty/Idris2.idr:
```diff
@@ -166,7 +166,7 @@ printStmts fl $ NewF ([< a, b ] ==> maybeRet) body cont = do
   rest <- printStmts @{NewFun nm {isInfix} @{names}} fl cont
   let infixAwareName : Doc opts = if isInfix then "(" <+> line nm <+> ")" else line nm
-  let processedInputTypes : Doc opts = hsep $ (snocListTyToList [< a, b ]) <&> (\ty => printTy ty <++> "->")
+  let processedInputTypes : Doc opts = hsep $ (toList [< a, b ]) <&> (\ty => printTy ty <++> "->")
```
Well, creating a snoc-list and immediately converting it to a list looks like we should create a list in the first place:

```suggestion
  let processedInputTypes : Doc opts = hsep $ ([a, b]) <&> (\ty => printTy ty <++> "->")
```

buzden commented on examples/pil-fun/src/Language/PilFun/Derived.idr:
```diff
@@ -8,5 +8,6 @@ import Deriving.DepTyCheck.Gen
 
 %logging "deptycheck.derive" 5
 %logging "deptycheck.derive.least-effort" 7
+%logging "deptycheck.util" 20
```
I thought -- why specialisation logging is in `deptycheck.util`, when this always happens only during derivation? Maybe, we should rename logging topic from `deptycheck.util.specialisation` to `deptycheck.derive.specialisation`, WDYT?

UARTman replied:
I'd also propose `deptycheck.specialisation` as a possible  alternative, but `deptycheck.derive.specialisation` sounds good to me as well.

buzden replied:
I'd stick to `deptycheck.derive.specialisation` because this enables a single-line `%logging deptycheck.derive <level>` to see important stuff when appropriate still enabling finer tuning when needed (like in this piece of code).

buzden commented on src/Deriving/DepTyCheck/Util/Specialisation.idr:
```diff
@@ -0,0 +1,307 @@
+-- Given a set of given argument indices, convert a list of their values into a vector that can be fed to callGen
+--
+-- The values should be listed for indices in ascending order
+-- (i.e. how said indices would be sorted if we called toList on the set)
```
```suggestion
||| Given a set of given argument indices, convert a list of their values into a vector that can be fed to `callGen`
|||
||| The values should be listed for indices in ascending order
||| (i.e. how these indices would be sorted if we called `toList` on the set)
```

UARTman committed:
[ cleanup ] Address more comments

- use MkUnit in SpecialiseData.idr tuple helpers
- Remove nonsensical snoc-to-list conversion in pil-fun Idris2 pretty-printer
- Flatten pattern matching in allQImpl
- Tweak formGivenVals docstring

UARTman committed:
[ refactor ] Change deptycheck specialisation logging

`deptycheck.util.specialisation` to `deptycheck.derive.specialisation`

UARTman committed:
[ pack ] Bump pack to 260827
